rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/geo/GeofencingProcessor.java

```diff
@@ -0,0 +1,198 @@
+    }
+
+    public List<EntityId> getOutsideGeofences(Long currentTs, Long durationMs, List<EntityId> matchedGeofences, Set<GeofenceState> geofenceStates) {
+        Set<GeofenceState> candidatesForOutside = geofenceStates.stream().filter(state -> state.isLeft() && !matchedGeofences.contains(state.getGeofenceId())).collect(Collectors.toSet());
+
+        if (CollectionUtils.isEmpty(candidatesForOutside)) {
+            return emptyList();
+        }
+
+        List<GeofenceState> outsideGeofences = new ArrayList<>();
+
+        for (GeofenceState candidate : candidatesForOutside) {
+            long diff = currentTs - candidate.getLeftTs();
+            boolean isTimeForOutside = diff > durationMs;
+            if (isTimeForOutside) {
+                outsideGeofences.add(candidate);
+            }
+        }
+
+        outsideGeofences.forEach(geofenceStates::remove);
+
+        return outsideGeofences.stream().map(GeofenceState::getGeofenceId).collect(Collectors.toList());
+    }
+
+    public boolean hasEnteredGeofence(List<EntityId> matchedGeofences, Set<GeofenceState> geofenceStates) {
+        return geofenceStates.stream().anyMatch(geofenceState -> geofenceState.isEntered() && matchedGeofences.contains(geofenceState.getGeofenceId()));
+    }
+
+    public boolean hasEnteredNotMatchedGeofence(List<EntityId> matchedGeofences, Set<GeofenceState> geofenceStates) {
+        return geofenceStates.stream().anyMatch(geofenceState -> geofenceState.isEntered() && !matchedGeofences.contains(geofenceState.getGeofenceId()));
+    }
+
+    public boolean hasNotEnteredAnyGeofence(Set<GeofenceState> geofenceStates) {
+        return geofenceStates.stream().allMatch(GeofenceState::isLeft);
+    }
+
+    private void persistState(EntityId entityId, RuleNodeId ruleNodeId, Set<GeofenceState> geofenceStates) {
+        ObjectNode jsonNode = JacksonUtil.newObjectNode();
+
+        jsonNode.set(GEOFENCE_STATES_KEY, JacksonUtil.valueToTree(geofenceStates));
+
+        BaseAttributeKvEntry attributeKvEntry = new BaseAttributeKvEntry(new JsonDataEntry(getGeofenceStateAttributeKey(ruleNodeId), jsonNode.toString()), System.currentTimeMillis());
+        context.getAttributesService().save(context.getTenantId(), entityId, AttributeScope.SERVER_SCOPE, attributeKvEntry);
+    }
+
+    private void deleteGeofenceStateByGeofenceId(EntityId geofenceId, Set<GeofenceState> geofenceStates) {
+        Optional<GeofenceState> geofenceState = geofenceStates.stream().filter(state -> state.getGeofenceId().equals(geofenceId)).findFirst();
+        geofenceState.ifPresent(geofenceStates::remove);
+    }
+
+    public Set<GeofenceState> fetchGeofenceStatesForEntity(EntityId entityId, RuleNodeId ruleNodeId) throws ExecutionException, InterruptedException {
+        Optional<AttributeKvEntry> attributeKvEntry = context.getAttributesService().find(context.getTenantId(), entityId, AttributeScope.SERVER_SCOPE, getGeofenceStateAttributeKey(ruleNodeId)).get();
```

Comment:
We can't do blocking operations .get() in the rule nodes since these block the rule node dispatcher threads from processing other nodes. You should use callbacks to handle future results. Please take a look at how we handle futures in other rule nodes.